Cc https and cors

kentaromiura opened the pull request:
Support for --https through a dummy autosigned SSL certificate (that expires in 2038 due to the unix 32bit millenium bug)
Support for --cors that adds some headers so one can testing ajax from a different domain/host/port (though its only for non njs files)

kentaromiura committed:
Adding https and cors features.

https is triggered by the --https command, works with every other
parameters
--cors allows cors for any Origin for any file apart for the njs one,
where one must manually add, for example:

  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Headers", "Cache-Control, Pragma,
Origin, Authorization, Content-Type, X-Requested-With");
  res.setHeader("Access-Control-Allow-Methods", "GET, PUT, POST");

kentaromiura committed:
Adding myself to the contributor list

..so when in 2038 the ssl certificate expires people can blame me,
instead of Andrea.

kentaromiura commented on build/polpetta:
```diff
@@ -2183,8 +2264,19 @@ if (!module.parent) {
+                  ca : Polpetta.prototype.ca,
+                  key: Polpetta.prototype.key
+              }, Polpetta.factory);
+              serverListen(server.on("error", serverListen.error));
```
we can move this down after the if

kentaromiura commented on build/polpetta:
```diff
@@ -2183,8 +2264,19 @@ if (!module.parent) {
+              serverListen(server.on("error", serverListen.error));
+          }else{
+              server = http.createServer(Polpetta.factory);
+              serverListen(server.on("error", serverListen.error));
```
...and remove this line

WebReflection replied:
please move `serverListen(server.on("error", serverListen.error));` after the `if/else` since it's common operation in both cases

kentaromiura commented on build/polpetta:
```diff
@@ -901,6 +917,14 @@ if (SEP == WEB_SEP) {
+if(CORS){
+    arguments.splice(arguments.indexOf('--cors'), 1)
+    root = polpetta_root(arguments)
+}
```
if you want you can simplify this by calling polpetta_root just once (removing the one in the root definition also), after these splice are done.

WebReflection replied:
what if both SSL and CORS are passed? _boom_ ?

kentaromiura replied:
It works just fine.

WebReflection replied:
that's not an answer. `polpetta_root` is based on arguments length and does operate expecting 1 maximum 2 arguments. Anything that is not `port` or `folder` should not be parsed for no reason through `path.resolve` so I would avoid multiple calls to `polpetta_root` when not necessary and here is not necessary.

Please move the `root = polpetta_root(arguments)` outside both `if/else` and put in eventually after either inside a `if(SSL || CORS)` condition or regardless, dropping the initial assignment on top being sure the rest of the following logic is not compromised.

Thanks

kentaromiura replied:
that's exactly what I meant with 

> if you want you can simplify this by calling polpetta_root just once (removing the one in the root definition also),  after these splice are done.

Right now it was eventually setting the root wrong a couple of time before fixing it, after the ifs the root is always correct, and that's why I told you that it works  just fine.

Anyway I'll make the couple of changes highlighted and then ping you again :)

WebReflection replied:
oh ... I see ... well, I am DRY code maniac so it would be ideal to have that call once since once should be needed to avoid any ambiguity or entry point to fail the bootstrap ^_^

Thanks

kentaromiura replied:
don't worry I removed both the if using short-circuiting logic while defining SSL and CORS, so root, HOST_USED and HOST_USER_PORT are not affected anymore by the two new params

kentaromiura committed:
cleanup a bit

using short-circuit to avoid using the if to fix root (and other
possibly argument dependent affected functions) moving the splice in
the SSL and CORS boolean definition (now they are always real boolean)

kentaromiura commented:
@WebReflection let me know what you want to do with the certificates

WebReflection commented:
uhm .. I might drop those doubled indexOf later on but overall looks OK. Certificates can stay there for now.

kentaromiura commented:
It should be simple to keep it inline also LIKE 
SSL= ~(SSL= indxof) && splice(SSL). . . 

Ps typing code from a phone is horrible.
